README.rst

```diff
@@ -1132,33 +1132,39 @@ achieves the move via a replication of the data objects found to the destination
 resource , followed by a trimming of each replica from the source.  We assume for our current
 purposed that all replicas are "good", ie have a status of "1" ::
 
+  import pprint, sys
   from irods.resource import iRODSResource
+  import irods.keywords as kw, irods.exception as ex, irods.test.helpers as h
   from irods.collection import iRODSCollection
   from irods.data_object import iRODSDataObject
   from irods.models import Resource,Collection,DataObject
-  def repl_and_trim (srcRescName, dstRescName = '', verbose = False):
+  def repl_and_trim (srcRescName, dstRescName = '', verbose = False, n_copies = 1, session = h.make_session()):
       objects_trimmed = 0
       q = session.query(Resource).filter(Resource.name == srcRescName)
       srcResc = iRODSResource( session.resources, q.one())
       # loop over data objects found on srcResc
       for q_row in session.query(Collection,DataObject) \
                           .filter(DataObject.resc_id == srcResc.id):
-          collection =  iRODSCollection (session.collections, result = q_row)
-          data_object = iRODSDataObject (session.data_objects, parent = collection, results = (q_row,))
-          objects_trimmed += 1
+          collection =  iRODSCollection(session.collections, result = q_row)
+          data_object = iRODSDataObject(session.data_objects, parent = collection, results = (q_row,))
           if verbose :
-              import pprint
               print( '--------', data_object.name, '--------')
               pprint.pprint( [vars(r) for r in data_object.replicas if
                               r.resource_name == srcRescName] )
           if dstRescName:
-              objects_trimmed += 1
               data_object.replicate(dstRescName)
-              for replica_number in [r.number for r in data_object.replicas]:
-                  options = { kw.DATA_REPL_KW: replica_number }
-                  data_object.unlink( **options )
+              repl_number_to_trim = [r.number for r in data_object.replicas if r.resource_name == srcRescName]
+              if repl_number_to_trim:
```

Comment:
Is checking `repl_number_to_trim` necessary?

Reply:
Technically no.  Because of the query being structured the way it is, as it falls out from the conditions of the query, it _should_ have exactly one element.  More if the version of iRODS is an older one in which a leaf resource could contain more than one.  Perhaps I should just loop over the array; if it has zero elements, that logic should take care of itself.

Reply:
On second thought that complicates logic too much.  I may have to rethink that section, if we still consider it an overriding concern...  or maybe leave out the `n_copies` parameter and just force it to 1.

Reply:
The code is fine to me. I was just highlighting that that if-statement seems unnecessary.

The function replicates and trims based on a source resource. That implies `repl_number_to_trim` is always valid.

If I'm missing something, please let me know.

Reply:
Well, `srcRescName` could (should?) be referring to a root resource in a hierarchy, in which case `repl_number_to_trim` would be invalid because data does not reside in coordinating resources. We would need to know from which leaf resource the object was replicating in the previous step in order to know which leaf resource to look for in the list of replicas. Not sure whether the replication API communicates this information or not.

Reply:
That's a good point.

The README should include some words about that or the example should be changed to use an explicit replica number.

Reply:
Does this script have any utility at all then?  Should we maybe start over,  or just accept much smaller examples demonstrating much less

Reply:
I don't think we need to change the examples. We just need to capture/explain the assumptions.

Reply:
or if we take Alans suggestion, I'll do the python equivalent of  `select ... where DATA_RESC_HIER like '<rootResc>;;%'`

Reply:
Hmm. Isn't all of this covered by L1145/1148?

Reply:
Good point. Well, the `resc_id` referred to on those lines ought to be a leaf resource, unless I'm totally missing the point. In that case, the resource wouldn't work either because the resource passed to the trim API must be a root resource, else you will encounter `DIRECT_CHILD_ACCESS`.

Reply:
> Good point. Well, the `resc_id` referred to on those lines ought to be a leaf resource, unless I'm totally missing the point. In that case, the resource wouldn't work either because the resource passed to the trim API must be a root resource, else you will encounter `DIRECT_CHILD_ACCESS`.

In my experience you can trim a replica by number without encountering  `DIRECT_CHILD_ACCESS`.  At any rate, I will investigate the options we have and/or document assumptions being made.

Reply:
Oh, I see. Yeah that seems fine. I know using replica number works but for some reason I thought we were passing the leaf resource to trim here. We are just using it to determine the replica number it seems